examples/c/profile.c

```diff
@@ -0,0 +1,209 @@
+	struct stacktrace_event *event = (struct stacktrace_event *)data;
+	int err;
+
+	if (size != sizeof(struct stacktrace_event)) {
+		fprintf(stderr, "Invalid size %ld != %ld\n", size, sizeof(struct stacktrace_event));
+		return 1;
+	}
+
+	printf("PID %d, CPU %d\n", event->pid, event->cpu_id);
+	ctx->cfgs[1].params.process.pid = event->pid;
+
+	if (event->kernel_stackid >= 0) {
+		err = bpf_map_lookup_elem(ctx->stackmap_fd, &event->kernel_stackid, kernel_stack);
+		if (err < 0) {
+			fprintf(stderr,
+				"Fail to lookup a stcktrace of kernel: errno=%d, stackid=%d\n",
+				err, event->kernel_stackid);
+			return 1;
+		}
+		printf("Kernel:\n");
+		show_stack_trace(ctx, (uint64_t *)kernel_stack);
+	} else
+		printf("No Kernel Stack\n");
+
+	if (event->user_stackid >= 0) {
+		err = bpf_map_lookup_elem(ctx->stackmap_fd, &event->user_stackid, user_stack);
+		if (err < 0) {
+			fprintf(stderr,
+				"Fail to lookup a stcktrace of userspace: errno=%d, stackid=%d\n",
+				err, event->user_stackid);
+			return 1;
+		}
+		printf("Userspace:\n");
+		show_stack_trace(ctx, (uint64_t *)user_stack);
+	} else
+		printf("No Userspace Stack\n");
+
+	printf("\n");
+	return 0;
+}
+
+int
+main(int argc, char * const argv[])
+{
+	int freq = 1, pid = -1, cpu;
+	struct profile_bpf *skel = NULL;
+	struct perf_event_attr attr;
+	struct bpf_link **links = NULL;
+	struct handler_ctx ctx;
+	struct ring_buffer *ring_buf = NULL;
+	int num_cpus = get_nprocs(); /* number of processors */
+	int *pefds = NULL, pefd;
+	int i, err;
+
+	memset(&ctx, 0, sizeof(ctx));
+
+	skel = profile_bpf__open();
+	if (!skel) {
+		fprintf(stderr, "Failed to open BPF skeleton\n");
+		return 1;
+	}
+
+	err = profile_bpf__load(skel);
+	if (err) {
+		fprintf(stderr, "Failed to load and verify BPF skeleton\n");
+		goto cleanup;
+	}
+
+	/* Data passing to event_handler */
+	ctx.stackmap_fd = bpf_map__fd(skel->maps.stackmap);
+	ctx.symbolizer = blazesym_new();
+	ctx.cfgs[0].cfg_type = CFG_T_KERNEL;
+	ctx.cfgs[1].cfg_type = CFG_T_PROCESS;
+	if (ctx.symbolizer == NULL) {
+		fprintf(stderr, "Fail to create a symbolizer\n");
+		err = -1;
+		goto cleanup;
+	}
+
+	/* Prepare ring buffer */
+	ring_buf = ring_buffer__new(bpf_map__fd(skel->maps.events),
+				    event_handler,
+				    &ctx,
+				    NULL);
+	if (ring_buf == NULL) {
+		err = -1;
+		goto cleanup;
+	}
+
+	pefds = (int *)malloc(sizeof(int) * num_cpus);
+	memset(pefds, -1, sizeof(int) * num_cpus);
+
+	links = (struct bpf_link **)malloc(sizeof(struct bpf_link *) * num_cpus);
+	memset(links, 0, sizeof(struct bpf_link *) * num_cpus);
+
+	memset(&attr, 0, sizeof(attr));
+	attr.type = PERF_TYPE_HARDWARE;
+	attr.size = sizeof(attr);
+	attr.config = PERF_COUNT_HW_CPU_CYCLES;
+	attr.sample_freq = freq;
+	attr.freq = 1;
+
+	for (cpu = 0; cpu < num_cpus; cpu++) {
+		/* Set up performance monitoring on a CPU/Core */
+		pefd = perf_event_open(&attr, pid, cpu, -1, PERF_FLAG_FD_CLOEXEC);
+		if (pefd < 0) {
+			perror("perf_event_open");
```

Comment:
stay consistent with how you report errors, you've been using fprintf(stderr) up till this point